[Style] Convert the 'font-*-settings' properties to strong style types

weinig opened the pull request:
#### ec5169a9d8fc1010ed60e51f7cf8c5f7d3014089
<pre>
[Style] Convert the &apos;font-*-settings&apos; properties to strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=300122">https://bugs.webkit.org/show_bug.cgi?id=300122</a>

Reviewed by Darin Adler.

Converts the &apos;font-feature-settings&apos; and &apos;font-variation-settings&apos;
properties to strong style types.

One test result progression due to now correct serialization of CSS number values.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/animations/font-variation-settings-interpolation-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/platform/graphics/FontCustomPlatformData.h:
* Source/WebCore/platform/graphics/FontTaggedSettings.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/style/StyleBuilderStateInlines.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationFunctions.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/StyleResolveForFont.cpp:
* Source/WebCore/style/StyleResolveForFont.h:
* Source/WebCore/style/values/fonts/StyleFontFeatureSettings.cpp: Added.
* Source/WebCore/style/values/fonts/StyleFontFeatureSettings.h: Added.
* Source/WebCore/style/values/fonts/StyleFontOpentypeTag.h: Added.
* Source/WebCore/style/values/fonts/StyleFontVariationSettings.cpp: Added.
* Source/WebCore/style/values/fonts/StyleFontVariationSettings.h: Added.

Canonical link: <a href="https://commits.webkit.org/301161@main">https://commits.webkit.org/301161@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ac0cec465c7e6660ea3d75d33dcf8c0dd99ab83c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125087 "12 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44755 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35493 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131935 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76960 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126964 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45446 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53320 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95222 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/63186 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128041 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36278 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111869 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75763 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35188 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30027 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75416 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106043 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30256 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134613 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51902 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39698 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103690 "Failed to checkout and rebase branch from PR 51766") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52332 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108084 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103463 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26348 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48806 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27099 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48955 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51791 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57581 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51161 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54517 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52852 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/animations/font-variation-settings-interpolation-expected.txt:
```diff
@@ -353,8 +353,8 @@ PASS Web Animations: property <font-variation-settings> from ['aaaa' 0, 'bbbb' 1
-FAIL CSS Transitions with transition: all: property <font-variation-settings> from ['aaaa' 30, 'bbbb' 20] to ['aaaa' 20, 'bbbb' 30] at (3.40282e+38) should be ['aaaa' -3.40282e+38, 'bbbb' 3.40282e+38] assert_array_equals: expected property 0 to be "\"aaaa\" -340282001837565597733306976381245063168" but got "\"aaaa\" -infinity" (expected array ["\"aaaa\" -340282001837565597733306976381245063168",  ...
-FAIL CSS Animations: property <font-variation-settings> from ['aaaa' 30, 'bbbb' 20] to ['aaaa' 20, 'bbbb' 30] at (3.40282e+38) should be ['aaaa' -3.40282e+38, 'bbbb' 3.40282e+38] assert_array_equals: expected property 0 to be "\"aaaa\" -340282001837565597733306976381245063168" but got "\"aaaa\" -infinity" (expected array ["\"aaaa\" -340282001837565597733306976381245063168", "\"bbbb\" 340282001837 ...
-FAIL Web Animations: property <font-variation-settings> from ['aaaa' 30, 'bbbb' 20] to ['aaaa' 20, 'bbbb' 30] at (3.40282e+38) should be ['aaaa' -3.40282e+38, 'bbbb' 3.40282e+38] assert_array_equals: expected property 0 to be "\"aaaa\" -340282001837565597733306976381245063168" but got "\"aaaa\" -infinity" (expected array ["\"aaaa\" -340282001837565597733306976381245063168", "\"bbbb\" 340282001837 ...
+FAIL CSS Transitions: property <font-variation-settings> from ['aaaa' 30, 'bbbb' 20] to ['aaaa' 20, 'bbbb' 30] at (3.40282e+38) should be ['aaaa' -3.40282e+38, 'bbbb' 3.40282e+38] assert_array_equals: expected property 0 to be "\"aaaa\" -340282001837565597733306976381245063168" but got "\"aaaa\" -340282346638528859811704183484516925440" (expected array ["\"aaaa\" -34028200183756559773330697638124 ...
+FAIL CSS Transitions with transition: all: property <font-variation-settings> from ['aaaa' 30, 'bbbb' 20] to ['aaaa' 20, 'bbbb' 30] at (3.40282e+38) should be ['aaaa' -3.40282e+38, 'bbbb' 3.40282e+38] assert_array_equals: expected property 0 to be "\"aaaa\" -340282001837565597733306976381245063168" but got "\"aaaa\" -340282346638528859811704183484516925440" (expected array ["\"aaaa\" -34028200183 ...
+FAIL CSS Animations: property <font-variation-settings> from ['aaaa' 30, 'bbbb' 20] to ['aaaa' 20, 'bbbb' 30] at (3.40282e+38) should be ['aaaa' -3.40282e+38, 'bbbb' 3.40282e+38] assert_array_equals: expected property 0 to be "\"aaaa\" -340282001837565597733306976381245063168" but got "\"aaaa\" -340282346638528859811704183484516925440" (expected array ["\"aaaa\" -340282001837565597733306976381245 ...
+FAIL Web Animations: property <font-variation-settings> from ['aaaa' 30, 'bbbb' 20] to ['aaaa' 20, 'bbbb' 30] at (3.40282e+38) should be ['aaaa' -3.40282e+38, 'bbbb' 3.40282e+38] assert_array_equals: expected property 0 to be "\"aaaa\" -340282001837565597733306976381245063168" but got "\"aaaa\" -340282346638528859811704183484516925440" (expected array ["\"aaaa\" -340282001837565597733306976381245 ...
```
Provocative test result; I wonder what we need to do to get this test passing.

weinig replied:
Pretty sure its just `FontVariationSettings<float>` should really be `FontVariationSettings<double>`.

darinadler commented on Source/WebCore/platform/graphics/FontCustomPlatformData.h:
```diff
@@ -55,8 +55,8 @@ class FontDescription;
-template <typename T> class FontTaggedSettings;
-typedef FontTaggedSettings<int> FontFeatureSettings;
+template<typename T> class FontTaggedSettings;
+using FontFeatureSettings = FontTaggedSettings<int>;
```
```suggestion
template<typename> class FontTaggedSettings;
using FontFeatureSettings = FontTaggedSettings<int>;
```

darinadler commented on Source/WebCore/style/values/fonts/StyleFontFeatureSettings.cpp:
```diff
@@ -0,0 +1,112 @@
+                return CSS::Keyword::Normal { };
+            state.setCurrentPropertyInvalidAtComputedValueTime();
+            return CSS::Keyword::Normal { };
+        }
```
```suggestion
        switch (auto valueID = primitiveValue->valueID(); valueID) {
        case CSSValueNormal:
            return CSS::Keyword::Normal { };
        default:
            if (CSSPropertyParserHelpers::isSystemFontShorthand(valueID))
                return CSS::Keyword::Normal { };
            state.setCurrentPropertyInvalidAtComputedValueTime();
            return CSS::Keyword::Normal { };
        }
```

darinadler commented on Source/WebCore/style/values/fonts/StyleFontVariationSettings.cpp:
```diff
@@ -0,0 +1,152 @@
+                return CSS::Keyword::Normal { };
+            state.setCurrentPropertyInvalidAtComputedValueTime();
+            return CSS::Keyword::Normal { };
+        }
```
```suggestion
        switch (auto valueID = primitiveValue->valueID(); valueID) {
        case CSSValueNormal:
            return CSS::Keyword::Normal { };
        default:
            if (CSSPropertyParserHelpers::isSystemFontShorthand(valueID))
                return CSS::Keyword::Normal { };
            state.setCurrentPropertyInvalidAtComputedValueTime();
            return CSS::Keyword::Normal { };
        }
```

weinig committed:
[Style] Convert the 'font-*-settings' properties to strong style types
https://bugs.webkit.org/show_bug.cgi?id=300122

Reviewed by Darin Adler.

Converts the 'font-feature-settings' and 'font-variation-settings'
properties to strong style types.

One test result progression due to now correct serialization of CSS number values.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/animations/font-variation-settings-interpolation-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/platform/graphics/FontCustomPlatformData.h:
* Source/WebCore/platform/graphics/FontTaggedSettings.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/style/StyleBuilderStateInlines.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationFunctions.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/StyleResolveForFont.cpp:
* Source/WebCore/style/StyleResolveForFont.h:
* Source/WebCore/style/values/fonts/StyleFontFeatureSettings.cpp: Added.
* Source/WebCore/style/values/fonts/StyleFontFeatureSettings.h: Added.
* Source/WebCore/style/values/fonts/StyleFontOpentypeTag.h: Added.
* Source/WebCore/style/values/fonts/StyleFontVariationSettings.cpp: Added.
* Source/WebCore/style/values/fonts/StyleFontVariationSettings.h: Added.

Canonical link: https://commits.webkit.org/301161@main
